# `pandas` - Clean DataFrame

Goal of this notebook is to create a useful dataset with appropriate values and variables.

__[EDIT]__

Deal with missing values

Create features (aka variables or columns)
- changing units
- splitting/joining strings

## Introduction

There are three basic steps to cleaning data:
1. Dealing with missing values
1. Manipulating strings 
1. Manipulating datetimes

Each is described in a separate section below.

First though libraries are loaded and define the `read_dataframe` function.

Load the libraries.

In [5]:
import pandas  as pd
import numpy   as np
(pd.__version__,
 np.__version__
)

Out[ 1 ]: ('0.18.1', '1.11.1')

Create a dataframe from the `imports-85.csv` CSV file. Set the column names.

In [7]:
def read_college_scorecard(file_path='/dbfs/mnt/datalab-datasets/college-scorecard/MERGED2013_PP.csv',
                           n_rows=None):
  return pd.read_csv(file_path,
                     nrows=n_rows,
                     na_values=['PrivacySuppressed']
                     )

In [8]:
read_college_scorecard().info()

/databricks/python/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2833: DtypeWarning: Columns (7,9) have mixed types. Specify dtype option on import or set low_memory=False.
 if self.run_code(code):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7804 entries, 0 to 7803
Columns: 1729 entries, UNITID to C200_4_POOLED_SUPP
dtypes: float64(1710), int64(12), object(7)
memory usage: 102.9+ MB

## Missing Values

Recall that the `isnull` method takes a Series/DataFrame as input and indicates whether the elements of the input are `NaN`. 

The output has the same shape as the input.

In [11]:
(read_college_scorecard().shape,
 read_college_scorecard().isnull().shape
)

Out[ 4 ]: ((7804, 1729), (7804, 1729))

The first five values of the `C150_L4_POOLED_SUPP` variable are missing.

In [13]:
read_college_scorecard().isnull().head()

Out[ 5 ]: 
 UNITID OPEID opeid6 INSTNM CITY STABBR ZIP AccredAgency INSTURL \
0 False False False False False False False False False 
1 False False False False False False False False False 
2 False False False False False False False False False 
3 False False False False False False False False False 
4 False False False False False False False False False 

 NPCURL ... PELL_RPY_3YR_RT_SUPP NOPELL_RPY_3YR_RT_SUPP \
0 False ... False False 
1 False ... False False 
2 False ... False False 
3 False ... False False 
4 False ... False False 

 FEMALE_RPY_3YR_RT_SUPP MALE_RPY_3YR_RT_SUPP FIRSTGEN_RPY_3YR_RT_SUPP \
0 False False False 
1 False False False 
2 False False False 
3 False False False 
4 False False False 

 NOTFIRSTGEN_RPY_3YR_RT_SUPP C150_L4_POOLED_SUPP C150_4_POOLED_SUPP \
0 False True False 
1 False True False 
2 False True True 
3 False True False 
4 False True False 

 C200_L4_POOLED_SUPP C200_4_POOLED_SUPP 
0 True False 
1 True False 
2 True True 
3 True False 
4 True False 

[5 rows x 1729 columns]

This can be used to list the variables/columns with missing values and indicate the number of missing values for each.

In [15]:
read_college_scorecard().isnull().sum(axis=0).sort_values(ascending=False)[lambda x: x > 0]

Out[ 6 ]: 
DEP_ENRL_ORIG_YR4_RT 7804
PELL_DEATH_YR4_RT 7804
MALE_WDRAW_2YR_TRANS_YR4_RT 7804
MALE_ENRL_ORIG_YR4_RT 7804
MALE_ENRL_4YR_TRANS_YR4_RT 7804
MALE_ENRL_2YR_TRANS_YR4_RT 7804
MALE_UNKN_ORIG_YR4_RT 7804
MALE_UNKN_4YR_TRANS_YR4_RT 7804
MALE_UNKN_2YR_TRANS_YR4_RT 7804
PELL_COMP_ORIG_YR4_RT 7804
PELL_UNKN_ORIG_YR4_RT 7804
PELL_COMP_4YR_TRANS_YR4_RT 7804
PELL_COMP_2YR_TRANS_YR4_RT 7804
PELL_WDRAW_ORIG_YR4_RT 7804
PELL_WDRAW_4YR_TRANS_YR4_RT 7804
PELL_WDRAW_2YR_TRANS_YR4_RT 7804
PELL_ENRL_ORIG_YR4_RT 7804
PELL_ENRL_4YR_TRANS_YR4_RT 7804
MALE_WDRAW_4YR_TRANS_YR4_RT 7804
MALE_WDRAW_ORIG_YR4_RT 7804
MALE_COMP_2YR_TRANS_YR4_RT 7804
MALE_COMP_4YR_TRANS_YR4_RT 7804
FEMALE_DEATH_YR4_RT 7804
FEMALE_COMP_ORIG_YR4_RT 7804
FEMALE_COMP_4YR_TRANS_YR4_RT 7804
FEMALE_COMP_2YR_TRANS_YR4_RT 7804
FEMALE_WDRAW_ORIG_YR4_RT 7804
FEMALE_WDRAW_4YR_TRANS_YR4_RT 7804
FEMALE_WDRAW_2YR_TRANS_YR4_RT 7804
FEMALE_ENRL_ORIG_YR4_RT 7804
 ... 
CIP39ASSOC 421
CIP39CERT2 421
CIP39CERT1 421
CIP27CERT1 421
CIP38CERT4 421
CIP38ASSOC 421
CIP38CERT2 421
CIP38CERT1 421
CIP31BACHL 421
CIP31CERT4 421
CIP31ASSOC 421
CIP38BACHL 421
CIP31CERT1 421
CIP29ASSOC 421
CIP30BACHL 421
CIP27CERT2 421
CIP27ASSOC 421
CIP27BACHL 421
CIP29CERT1 421
CIP29CERT2 421
CIP27CERT4 421
CIP29CERT4 421
CIP29BACHL 421
CIP30CERT1 421
CIP30CERT2 421
CIP30ASSOC 421
CIP30CERT4 421
APPL_SCH_N 324
INSTURL 85
CDR3 62
dtype: int64

This list has many elements so a visual inspection isn't very useful. 

The `describe` method summarizes the number of missing values of the variables.

In [17]:
read_college_scorecard().isnull().sum(axis=0).sort_values(ascending=False)[lambda x: x > 0].describe()

Out[ 7 ]: 
count 1713.000000
mean 6044.914186
std 2861.623608
min 62.000000
25% 4189.000000
50% 7804.000000
75% 7804.000000
max 7804.000000
dtype: float64

Notice that over half of the variables contain no valid (non-missing) values. 

In the next sections:
- Columns will be dropped that do not meet a certain threshold of valid values.
- Columns will be dropped that do not meet a certain threshold of valid values.
- The remaining missing values will be imputed (filled in.)

### Drop Rows/Columns (`dropna` Method)
- https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.dropna.html

Notice the number of rows and columns in the dataframe.

In [21]:
read_college_scorecard()\
  .shape

Out[ 8 ]: (7804, 1729)

Drop all columns with fewer than 90% (7024) valid values,

In [23]:
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .shape

Out[ 9 ]: (7804, 281)

List the columns that remain.

In [25]:
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .pipe(lambda df: list(df.columns))

Out[ 10 ]: 
['\ufeffUNITID',
 'OPEID',
 'opeid6',
 'INSTNM',
 'CITY',
 'STABBR',
 'ZIP',
 'INSTURL',
 'NPCURL',
 'HCM2',
 'main',
 'NUMBRANCH',
 'PREDDEG',
 'HIGHDEG',
 'CONTROL',
 'st_fips',
 'region',
 'LOCALE',
 'HBCU',
 'PBI',
 'ANNHI',
 'TRIBAL',
 'AANAPII',
 'HSI',
 'NANTI',
 'MENONLY',
 'WOMENONLY',
 'PCIP01',
 'PCIP03',
 'PCIP04',
 'PCIP05',
 'PCIP09',
 'PCIP10',
 'PCIP11',
 'PCIP12',
 'PCIP13',
 'PCIP14',
 'PCIP15',
 'PCIP16',
 'PCIP19',
 'PCIP22',
 'PCIP23',
 'PCIP24',
 'PCIP25',
 'PCIP26',
 'PCIP27',
 'PCIP29',
 'PCIP30',
 'PCIP31',
 'PCIP38',
 'PCIP39',
 'PCIP40',
 'PCIP41',
 'PCIP42',
 'PCIP43',
 'PCIP44',
 'PCIP45',
 'PCIP46',
 'PCIP47',
 'PCIP48',
 'PCIP49',
 'PCIP50',
 'PCIP51',
 'PCIP52',
 'PCIP54',
 'CIP01CERT1',
 'CIP01CERT2',
 'CIP01ASSOC',
 'CIP01CERT4',
 'CIP01BACHL',
 'CIP03CERT1',
 'CIP03CERT2',
 'CIP03ASSOC',
 'CIP03CERT4',
 'CIP03BACHL',
 'CIP04CERT1',
 'CIP04CERT2',
 'CIP04ASSOC',
 'CIP04CERT4',
 'CIP04BACHL',
 'CIP05CERT1',
 'CIP05CERT2',
 'CIP05ASSOC',
 'CIP05CERT4',
 'CIP05BACHL',
 'CIP09CERT1',
 'CIP09CERT2',
 'CIP09ASSOC',
 'CIP09CERT4',
 'CIP09BACHL',
 'CIP10CERT1',
 'CIP10CERT2',
 'CIP10ASSOC',
 'CIP10CERT4',
 'CIP10BACHL',
 'CIP11CERT1',
 'CIP11CERT2',
 'CIP11ASSOC',
 'CIP11CERT4',
 'CIP11BACHL',
 'CIP12CERT1',
 'CIP12CERT2',
 'CIP12ASSOC',
 'CIP12CERT4',
 'CIP12BACHL',
 'CIP13CERT1',
 'CIP13CERT2',
 'CIP13ASSOC',
 'CIP13CERT4',
 'CIP13BACHL',
 'CIP14CERT1',
 'CIP14CERT2',
 'CIP14ASSOC',
 'CIP14CERT4',
 'CIP14BACHL',
 'CIP15CERT1',
 'CIP15CERT2',
 'CIP15ASSOC',
 'CIP15CERT4',
 'CIP15BACHL',
 'CIP16CERT1',
 'CIP16CERT2',
 'CIP16ASSOC',
 'CIP16CERT4',
 'CIP16BACHL',
 'CIP19CERT1',
 'CIP19CERT2',
 'CIP19ASSOC',
 'CIP19CERT4',
 'CIP19BACHL',
 'CIP22CERT1',
 'CIP22CERT2',
 'CIP22ASSOC',
 'CIP22CERT4',
 'CIP22BACHL',
 'CIP23CERT1',
 'CIP23CERT2',
 'CIP23ASSOC',
 'CIP23CERT4',
 'CIP23BACHL',
 'CIP24CERT1',
 'CIP24CERT2',
 'CIP24ASSOC',
 'CIP24CERT4',
 'CIP24BACHL',
 'CIP25CERT1',
 'CIP25CERT2',
 'CIP25ASSOC',
 'CIP25CERT4',
 'CIP25BACHL',
 'CIP26CERT1',
 'CIP26CERT2',
 'CIP26ASSOC',
 'CIP26CERT4',
 'CIP26BACHL',
 'CIP27CERT1',
 'CIP27CERT2',
 'CIP27ASSOC',
 'CIP27CERT4',
 'CIP27BACHL',
 'CIP29CERT1',
 'CIP29CERT2',
 'CIP29ASSOC',
 'CIP29CERT4',
 'CIP29BACHL',
 'CIP30CERT1',
 'CIP30CERT2',
 'CIP30ASSOC',
 'CIP30CERT4',
 'CIP30BACHL',
 'CIP31CERT1',
 'CIP31CERT2',
 'CIP31ASSOC',
 'CIP31CERT4',
 'CIP31BACHL',
 'CIP38CERT1',
 'CIP38CERT2',
 'CIP38ASSOC',
 'CIP38CERT4',
 'CIP38BACHL',
 'CIP39CERT1',
 'CIP39CERT2',
 'CIP39ASSOC',
 'CIP39CERT4',
 'CIP39BACHL',
 'CIP40CERT1',
 'CIP40CERT2',
 'CIP40ASSOC',
 'CIP40CERT4',
 'CIP40BACHL',
 'CIP41CERT1',
 'CIP41CERT2',
 'CIP41ASSOC',
 'CIP41CERT4',
 'CIP41BACHL',
 'CIP42CERT1',
 'CIP42CERT2',
 'CIP42ASSOC',
 'CIP42CERT4',
 'CIP42BACHL',
 'CIP43CERT1',
 'CIP43CERT2',
 'CIP43ASSOC',
 'CIP43CERT4',
 'CIP43BACHL',
 'CIP44CERT1',
 'CIP44CERT2',
 'CIP44ASSOC',
 'CIP44CERT4',
 'CIP44BACHL',
 'CIP45CERT1',
 'CIP45CERT2',
 'CIP45ASSOC',
 'CIP45CERT4',
 'CIP45BACHL',
 'CIP46CERT1',
 'CIP46CERT2',
 'CIP46ASSOC',
 'CIP46CERT4',
 'CIP46BACHL',
 'CIP47CERT1',
 'CIP47CERT2',
 'CIP47ASSOC',
 'CIP47CERT4',
 'CIP47BACHL',
 'CIP48CERT1',
 'CIP48CERT2',
 'CIP48ASSOC',
 'CIP48CERT4',
 'CIP48BACHL',
 'CIP49CERT1',
 'CIP49CERT2',
 'CIP49ASSOC',
 'CIP49CERT4',
 'CIP49BACHL',
 'CIP50CERT1',
 'CIP50CERT2',
 'CIP50ASSOC',
 'CIP50CERT4',
 'CIP50BACHL',
 'CIP51CERT1',
 'CIP51CERT2',
 'CIP51ASSOC',
 'CIP51CERT4',
 'CIP51BACHL',
 'CIP52CERT1',
 'CIP52CERT2',
 'CIP52ASSOC',
 'CIP52CERT4',
 'CIP52BACHL',
 'CIP54CERT1',
 'CIP54CERT2',
 'CIP54ASSOC',
 'CIP54CERT4',
 'CIP54BACHL',
 'DISTANCEONLY',
 'UGDS',
 'UGDS_WHITE',
 'UGDS_BLACK',
 'UGDS_HISP',
 'UGDS_ASIAN',
 'UGDS_AIAN',
 'UGDS_NHPI',
 'UGDS_2MOR',
 'UGDS_NRA',
 'UGDS_UNKN',
 'PPTUG_EF',
 'CURROPER',
 'TUITFTE',
 'INEXPFTE',
 'PCTPELL',
 'PCTFLOAN',
 'UG25abv',
 'CDR3',
 'DEP_STAT_PCT_IND',
 'APPL_SCH_PCT_GE2',
 'DEP_INC_AVG',
 'IND_INC_AVG',
 'DEP_INC_N',
 'IND_INC_N',
 'APPL_SCH_N']

Drop all rows with fewer than 90% (254) valid values.

In [27]:
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .dropna(thresh= 254,axis=0)\
  .shape

Out[ 11 ]: (7086, 281)

Notice that there are now `7086` rows after dropping rows with fewer than `254` valid values. (We had `7804` rows.)

List the number of remaining missing values for all variables. In all cases we should have fewer than 10% (780) missing values.

In [30]:
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .dropna(thresh= 254,axis=0)\
  .isnull()\
  .sum(axis=0)\
  .sort_values(ascending=False)[lambda x: x > 0]

Out[ 12 ]: 
IND_INC_AVG 444
DEP_INC_AVG 433
IND_INC_N 426
DEP_INC_N 426
DEP_STAT_PCT_IND 355
APPL_SCH_PCT_GE2 249
APPL_SCH_N 73
NPCURL 64
UG25abv 61
CDR3 59
INSTURL 57
PCTPELL 23
PCTFLOAN 23
TUITFTE 21
INEXPFTE 21
PPTUG_EF 14
UGDS_HISP 3
UGDS 3
UGDS_WHITE 3
UGDS_BLACK 3
UGDS_ASIAN 3
UGDS_UNKN 3
UGDS_NRA 3
UGDS_AIAN 3
LOCALE 3
UGDS_2MOR 3
UGDS_NHPI 3
dtype: int64

### Replace Missing Values (`fillna` Method)

- https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.fillna.html

Filling forward/backward, which is often used to fill missing values in time series, is not covered here.

Which variable should not be imputed? The default imputation method is the median of the valid values. Consider whether there are variables  that:
1. require a different imputation method
1. should never be missing or imputed

The second case implies that rows with missing values for that variable should be dropped.

Check below for variables with missing values (to impute.)

In [35]:
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .dropna(thresh= 254,axis=0)\
  .isnull()\
  .sum(axis=0)\
  .sort_values(ascending=False)[lambda x: x > 0]

Out[ 13 ]: 
IND_INC_AVG 444
DEP_INC_AVG 433
IND_INC_N 426
DEP_INC_N 426
DEP_STAT_PCT_IND 355
APPL_SCH_PCT_GE2 249
APPL_SCH_N 73
NPCURL 64
UG25abv 61
CDR3 59
INSTURL 57
PCTPELL 23
PCTFLOAN 23
TUITFTE 21
INEXPFTE 21
PPTUG_EF 14
UGDS_HISP 3
UGDS 3
UGDS_WHITE 3
UGDS_BLACK 3
UGDS_ASIAN 3
UGDS_UNKN 3
UGDS_NRA 3
UGDS_AIAN 3
LOCALE 3
UGDS_2MOR 3
UGDS_NHPI 3
dtype: int64

Check the datatypes of the columns with missing values.

In [37]:
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .dropna(thresh= 254,axis=0)\
  .pipe(lambda df: df.loc[:,df.isnull().any()])\
  .dtypes

Out[ 14 ]: 
INSTURL object
NPCURL object
LOCALE float64
UGDS float64
UGDS_WHITE float64
UGDS_BLACK float64
UGDS_HISP float64
UGDS_ASIAN float64
UGDS_AIAN float64
UGDS_NHPI float64
UGDS_2MOR float64
UGDS_NRA float64
UGDS_UNKN float64
PPTUG_EF float64
TUITFTE float64
INEXPFTE float64
PCTPELL float64
PCTFLOAN float64
UG25abv float64
CDR3 float64
DEP_STAT_PCT_IND float64
APPL_SCH_PCT_GE2 float64
DEP_INC_AVG float64
IND_INC_AVG float64
DEP_INC_N float64
IND_INC_N float64
APPL_SCH_N float64
dtype: object

As a first attempt I tried using values as 'median'. This silently didn't work. Notice below.

In [39]:
NPCURL_ndx = read_college_scorecard()\
               .dropna(thresh=7024,axis=1)\
               .dropna(thresh= 254,axis=0)\
               .loc[:,'NPCURL']\
               .isnull()
NPCURL_ndx

/databricks/python/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2827: DtypeWarning: Columns (7,9) have mixed types. Specify dtype option on import or set low_memory=False.
 if self.run_code(code):
 Out[ 15 ]: 
0 False
1 False
2 False
3 False
4 False
5 False
6 False
7 False
8 False
9 False
10 False
11 False
12 False
13 False
14 False
15 False
16 False
17 False
18 False
19 False
20 False
21 False
22 False
23 False
24 False
25 False
26 False
27 False
28 False
29 False
 ... 
7353 False
7354 False
7355 False
7356 False
7357 False
7358 False
7359 False
7360 False
7361 False
7362 False
7363 False
7364 False
7365 False
7366 False
7367 False
7368 False
7369 False
7370 False
7371 False
7372 False
7373 False
7374 False
7375 False
7376 False
7377 False
7378 False
7379 False
7380 False
7381 False
7382 False
Name: NPCURL, dtype: bool

In [40]:
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .dropna(thresh= 254,axis=0)\
  .fillna(value='median')\
  .loc[NPCURL_ndx,['NPCURL','UGDS']]

Out[ 16 ]: 
 NPCURL UGDS
58 median 80
73 median 2
165 median 76
387 median 1101
397 median 182
461 median 191
494 median 506
691 median 422
895 median 53
918 median 60
990 median 57
1063 median median
1149 median 121
1152 median 124
1164 median 220
1614 median 738
1880 median 200
1918 median 0
1932 median 131
2069 median 654
2114 median 281
2130 median 375
2196 median 810
2207 median 8
2262 median 2766
2274 median 41
2430 median 79
2530 median 15
2647 median 346
3013 median 99
... ... ...
3436 median 664
3449 median 25
3517 median 205
3564 median 18
3698 median 243
3778 median 310
3840 median 10
3866 median 701
4246 median 88
4270 median 650
4338 median 25
4340 median 137
4375 median 28
4583 median 108
4758 median 20
4769 median 72
4972 median 212
5010 median 313
5244 median 6
5265 median 91
5431 median 31
5646 median 6
6181 median 22
6358 median 119
6404 median 38
6511 median 126
6585 median 60
6760 median 118
6836 median 249
7278 median 165

[64 rows x 2 columns]

Replacing the `object` variables `NPCURL` and `INSTURL` with the string `NA` seems reasonable, replacing it with the mode does not seem reasonable. The code below does just that.

In [42]:
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .dropna(thresh= 254,axis=0)\
  .fillna(value={'NPCURL': 'NA', 
                 'INSTURL': 'NA'})\
  .loc[NPCURL_ndx,['NPCURL','UGDS']]

Out[ 17 ]: 
 NPCURL UGDS
58 NA 80.0
73 NA 2.0
165 NA 76.0
387 NA 1101.0
397 NA 182.0
461 NA 191.0
494 NA 506.0
691 NA 422.0
895 NA 53.0
918 NA 60.0
990 NA 57.0
1063 NA NaN
1149 NA 121.0
1152 NA 124.0
1164 NA 220.0
1614 NA 738.0
1880 NA 200.0
1918 NA 0.0
1932 NA 131.0
2069 NA 654.0
2114 NA 281.0
2130 NA 375.0
2196 NA 810.0
2207 NA 8.0
2262 NA 2766.0
2274 NA 41.0
2430 NA 79.0
2530 NA 15.0
2647 NA 346.0
3013 NA 99.0
... ... ...
3436 NA 664.0
3449 NA 25.0
3517 NA 205.0
3564 NA 18.0
3698 NA 243.0
3778 NA 310.0
3840 NA 10.0
3866 NA 701.0
4246 NA 88.0
4270 NA 650.0
4338 NA 25.0
4340 NA 137.0
4375 NA 28.0
4583 NA 108.0
4758 NA 20.0
4769 NA 72.0
4972 NA 212.0
5010 NA 313.0
5244 NA 6.0
5265 NA 91.0
5431 NA 31.0
5646 NA 6.0
6181 NA 22.0
6358 NA 119.0
6404 NA 38.0
6511 NA 126.0
6585 NA 60.0
6760 NA 118.0
6836 NA 249.0
7278 NA 165.0

[64 rows x 2 columns]

The remaining variables are numeric and will be filled by their (respective) medians, but each variable has a different median.

This command creates a dictionary whose keys are variables with missing values and values are the medians (of the key variable.) These medians are calculated by ignoring the missing values.

In [44]:
median_dict = \
  read_college_scorecard()\
    .dropna(thresh=7024,axis=1)\
    .dropna(thresh= 254,axis=0)\
    .fillna(value={'NPCURL': 'NA', 
                   'INSTURL': 'NA'})\
    .pipe(lambda df: df.loc[:,df.isnull().any()])\
    .apply(lambda x: x.median(skipna=True),axis=0)\
    .to_dict()
(median_dict['UGDS'], 
 median_dict
)

Out[ 18 ]: 
(434.0,
 {'UGDS_ASIAN': 0.011900000000000001,
 'PCTFLOAN': 0.59260000000000002,
 'CDR3': 0.127,
 'UGDS_NRA': 0.0,
 'UGDS_WHITE': 0.56520000000000004,
 'PPTUG_EF': 0.15034999999999998,
 'DEP_STAT_PCT_IND': 0.59765625,
 'DEP_INC_AVG': 41768.517800000001,
 'PCTPELL': 0.51649999999999996,
 'UGDS_UNKN': 0.0146,
 'UG25abv': 0.41339999999999999,
 'IND_INC_N': 452.5,
 'UGDS': 434.0,
 'LOCALE': 21.0,
 'DEP_INC_N': 534.0,
 'UGDS_NHPI': 0.0,
 'UGDS_HISP': 0.068500000000000005,
 'TUITFTE': 8595.0,
 'INEXPFTE': 5085.0,
 'UGDS_2MOR': 0.0155,
 'APPL_SCH_PCT_GE2': 0.41740268200000002,
 'UGDS_BLACK': 0.099400000000000002,
 'IND_INC_AVG': 18127.085595,
 'UGDS_AIAN': 0.0027000000000000001,
 'APPL_SCH_N': 981.0})

The `median_dict` dictionary is used as the `value` parameter below.

In [46]:
median_dict = \
  read_college_scorecard()\
    .dropna(thresh=7024,axis=1)\
    .dropna(thresh= 254,axis=0)\
    .fillna(value={'NPCURL': 'NA', 
                   'INSTURL': 'NA'})\
    .pipe(lambda df: df.loc[:,df.isnull().any()])\
    .apply(lambda x: x.median(skipna=True),axis=0)\
    .to_dict()
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .dropna(thresh= 254,axis=0)\
  .fillna(value={'NPCURL': 'NA',
                 'INSTURL': 'NA'})\
  .fillna(value=median_dict)\
  .loc[NPCURL_ndx,['NPCURL','UGDS']]

Out[ 19 ]: 
 NPCURL UGDS
58 NA 80.0
73 NA 2.0
165 NA 76.0
387 NA 1101.0
397 NA 182.0
461 NA 191.0
494 NA 506.0
691 NA 422.0
895 NA 53.0
918 NA 60.0
990 NA 57.0
1063 NA 434.0
1149 NA 121.0
1152 NA 124.0
1164 NA 220.0
1614 NA 738.0
1880 NA 200.0
1918 NA 0.0
1932 NA 131.0
2069 NA 654.0
2114 NA 281.0
2130 NA 375.0
2196 NA 810.0
2207 NA 8.0
2262 NA 2766.0
2274 NA 41.0
2430 NA 79.0
2530 NA 15.0
2647 NA 346.0
3013 NA 99.0
... ... ...
3436 NA 664.0
3449 NA 25.0
3517 NA 205.0
3564 NA 18.0
3698 NA 243.0
3778 NA 310.0
3840 NA 10.0
3866 NA 701.0
4246 NA 88.0
4270 NA 650.0
4338 NA 25.0
4340 NA 137.0
4375 NA 28.0
4583 NA 108.0
4758 NA 20.0
4769 NA 72.0
4972 NA 212.0
5010 NA 313.0
5244 NA 6.0
5265 NA 91.0
5431 NA 31.0
5646 NA 6.0
6181 NA 22.0
6358 NA 119.0
6404 NA 38.0
6511 NA 126.0
6585 NA 60.0
6760 NA 118.0
6836 NA 249.0
7278 NA 165.0

[64 rows x 2 columns]

Check that there are no remaining missing values.

In [48]:
median_dict = \
  read_college_scorecard()\
    .dropna(thresh=7024,axis=1)\
    .dropna(thresh= 254,axis=0)\
    .fillna(value={'NPCURL': 'NA', 
                   'INSTURL': 'NA'})\
    .pipe(lambda df: df.loc[:,df.isnull().any()])\
    .apply(lambda x: x.median(skipna=True),axis=0)\
    .to_dict()
read_college_scorecard()\
  .dropna(thresh=7024,axis=1)\
  .dropna(thresh= 254,axis=0)\
  .fillna(value={'NPCURL': 'NA',
                 'INSTURL': 'NA'})\
  .fillna(value=median_dict)\
  .isnull()\
  .sum(axis=0)\
  .sort_values(ascending=False)

Out[ 20 ]: 
APPL_SCH_N 0
CIP11CERT1 0
CIP09BACHL 0
CIP10CERT1 0
CIP10CERT2 0
CIP10ASSOC 0
CIP10CERT4 0
CIP10BACHL 0
CIP11CERT2 0
CIP09ASSOC 0
CIP11ASSOC 0
CIP11CERT4 0
CIP11BACHL 0
CIP12CERT1 0
CIP12CERT2 0
CIP12ASSOC 0
CIP09CERT4 0
CIP09CERT2 0
CIP23BACHL 0
CIP04ASSOC 0
CIP03CERT2 0
CIP03ASSOC 0
CIP03CERT4 0
CIP03BACHL 0
CIP04CERT1 0
CIP04CERT2 0
CIP04CERT4 0
CIP09CERT1 0
CIP04BACHL 0
CIP05CERT1 0
 ..
CIP42CERT2 0
CIP42ASSOC 0
CIP42CERT4 0
CIP42BACHL 0
CIP43CERT2 0
CIP38CERT2 0
CIP43ASSOC 0
CIP43CERT4 0
CIP43BACHL 0
CIP44CERT1 0
CIP44CERT2 0
CIP44ASSOC 0
CIP41CERT4 0
CIP41ASSOC 0
CIP41CERT2 0
CIP41CERT1 0
CIP40BACHL 0
CIP40CERT4 0
CIP40ASSOC 0
CIP40CERT2 0
CIP40CERT1 0
CIP39BACHL 0
CIP39CERT4 0
CIP39ASSOC 0
CIP39CERT2 0
CIP39CERT1 0
CIP38BACHL 0
CIP38CERT4 0
CIP38ASSOC 0
UNITID 0
dtype: int64

## Final Code

The code that reads the dataset and deals with missing values is contained in the cell below.

In [50]:
def read_college_scorecard(file_path='/dbfs/mnt/datalab-datasets/college-scorecard/MERGED2013_PP.csv',
                           n_rows=None):
  return pd.read_csv(file_path,
                     nrows=n_rows,
                     na_values=['PrivacySuppressed']
                     )
def fix_missing_values(df):
  median_dict = \
    read_college_scorecard()\
      .dropna(thresh=7024,axis=1)\
      .dropna(thresh= 254,axis=0)\
      .fillna(value={'NPCURL': 'NA', 
                     'INSTURL': 'NA'})\
      .pipe(lambda df: df.loc[:,df.isnull().any()])\
      .apply(lambda x: x.median(skipna=True),axis=0)\
      .to_dict()
  return df.dropna(thresh=7024,axis=1)\
           .dropna(thresh= 254,axis=0)\
           .fillna(value={'NPCURL': 'NA',
                          'INSTURL': 'NA'})\
           .fillna(value=median_dict)

These functions are used as:

In [52]:
read_college_scorecard()\
  .pipe(fix_missing_values)

There's always something. See warning below.

In [54]:
x = read_college_scorecard().pipe(fix_missing_values)

In [55]:
x.columns[[7,9]]

Out[ 32 ]: Index(['INSTURL', 'HCM2'], dtype='object')

In [56]:
x.iloc[:,[7,9]].dtypes

Out[ 29 ]: 
INSTURL object
HCM2 int64
dtype: object

In [57]:
x.iloc[:,[7,9]].head()

Out[ 33 ]: 
 INSTURL HCM2
0 www.aamu.edu/ 0
1 www.uab.edu 0
2 www.amridgeuniversity.edu 0
3 www.uah.edu 0
4 www.alasu.edu/email/index.aspx 0

__The End__